ambry-store/src/main/java/com/github/ambry/store/IndexSegment.java

```diff
@@ -460,6 +460,29 @@ NavigableSet<IndexValue> find(StoreKey keyToFind) throws StoreException {
     return toReturn != null ? Collections.unmodifiableNavigableSet(toReturn) : null;
   }
 
+  /**
+   * Return all the keys in the IndexSegment in a set.
+   * @return
+   */
+  Set<StoreKey> getAllStoreKeys() throws StoreException {
+    Set<StoreKey> keys = new HashSet<>();
+    rwLock.readLock().lock();
+    boolean isSealed = sealed.get();
+    NavigableMap<StoreKey, ConcurrentSkipListSet<IndexValue>> indexCopy = index;
```

Comment:
it's fine, in the end, we don't really copy the keys in unsealed index segment.

Reply:
So which means we could use index directly right? Just don't get why do we need to copy the reference?  The indexCopy would have same value as index since it's just copy the reference.

Reply:
the index might be set to null due to the index segment becoming sealed. So in the lock, we have to obtain a copy of reference to the index.